pydarn/pydmap/io.py

```diff
@@ -0,0 +1,1296 @@
+        """
+        array_name = self.read_data('s', 1)
+        array_type = self.read_data('c', 1)
+
+        self.check_data_type(array_type, array_name)
+
+        array_type_fmt = DMAP_DATA_TYPES[array_type][0]
+        array_fmt_bytes = DMAP_DATA_TYPES[array_type][1]
+
+        array_dimension = self.read_data('i', 4)
+        self.bytes_check(array_dimension, "array dimension",
+                         record_size, "record size")
+        self.zero_check(array_dimension, "array dimension")
+
+        array_shape = [self.read_data('i', 4) for i in range(0, array_dimension)]
+        array_shape.reverse()
+
+        # if shape list is empty
+        if len(array_shape) != array_dimension:
+            message = "Error: Array shape {shape} could not be read."\
+                    " cursor: {cursor}".format(shape=array_shape,
+                                               cursor=self.cursor)
+            raise pydmap_exceptions.DmapDataError(self.dmap_file, message)
+        # TODO Explain where does slist come from?
+        elif any(x <= 0 for x in array_shape) and array_name != "slist":
+            message = "Error: Array shape {shape} contains "\
+                    "dimension size <= 0."\
+                    " Cursor: {cursor}".format(dim=array_shape,
+                                               cursor=self.cursor)
+            raise pydmap_exceptions.DmapDataError(self.filename, message)
+
+        for i in range(array_dimension):
+            if array_shape[i] >= record_size:
+                message = "Error: Array {index}-dimension size {size}"\
+                        " exceeds record size: {rec_size}."\
+                        "".format(index=i,
+                                  size=array_shape[i],
+                                  rec_size=record_size)
+                raise pydmap_exceptions.DmapDataError(self.dmap_file, message)
+
+        # We could use np.prod(array_shape) but the for loop has a better
+        # time performance.
+        total_num_cells = 1
+        for i in array_shape:
+            total_num_cells *= i
+        self.bytes_check(total_num_cells, "total number of cells",
+                         record_size, "record size")
+
+        total_num_cells_bytes = total_num_cells * array_fmt_bytes
+        self.bytes_check(total_num_cells_bytes,
+                         "total number of cells in bytes",
+                         record_size, "record size")
+
+        # parsing an array of strings requires a different method. Numpy can't
+        # parse strings or dmaps into arrays the way it can for other
+        # types because it doesn't
+        # know the sizes. They have to be manually read the slow way.
+        # Because chars
+        # are encoded as hex literals, they have to be read one at a
+        # time to make sense.
+
+        # what is DMAP format? Can we get an example that has a
+        # list of DMAPs?
+        # Also why is char also included?
```

Comment:
I think again since they are hex literals, but I can't recall exactly.

Reply:
Legacy comments. 

But I forgot to check for DMAP types to raise an error, Fixed!